subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchain.java

```diff
@@ -48,7 +48,7 @@ public JavaToolchain(JvmInstallationMetadata metadata, JavaCompilerFactory compi
         this.javaVersion = JavaLanguageVersion.of(metadata.getLanguageVersion().getMajorVersion());
         this.compilerFactory = compilerFactory;
         this.toolFactory = toolFactory;
-        this.implementationVersion = VersionNumber.parse(metadata.getImplementationVersion());
+        this.implementationVersion = VersionNumber.withPatchNumber().parse(metadata.getImplementationVersion());
```

Comment:
That change makes sense 👍 

Following our discussion on this topic though, we still have potential issues depending on how versions from OpenJDK like `11.0.9+11.1` and `11.0.9+11` are represented in `java.version` because anything starting with that `+` is ignored, right?
I am not saying this needs to be fixed in that PR to be clear.

Reply:
According to the [JEP 223](https://openjdk.java.net/jeps/223) the build number (the part after '+') isn't included in the `java.version` and, therefore, in the result of `getImplementationVersion`, so yes, it is effectively ignored. 

It seems that my issue with JDK 8 was only the tip of the iceberg. I did some more research on the topic, this time checking several JDK vendors. First, it is possible to have builds to differ only in a build number; these are [issued weekly](https://wiki.openjdk.java.net/display/JDKUpdates/Detailed+Process+Description) in the OpenJDK project. These builds are more like nightlies, so they aren't widely used. Only the last one is promoted to the General Availability. For the developers who use nightly OpenJDK builds it might make sense to look into build number for toolchain selection as well. Second, nothing prevents vendors from shipping a new downstream build while keeping the full version string. The Amazon Corretto does this for bugfix releases. For example, they have `11.0.6.10.1` and `11.0.6.10.1-1`, both based on OpenJDK `11.0.6`. The full Corretto version is stored in  `java.vendor.version` while `java.version` seems to be taken from upstream as is.

Out of curiosity, I dug into the `+11.1` story a bit; looks like AdoptOpenJDK rebuilt the version to [include the ShenandoahGC](https://github.com/adoptium/temurin-build/issues/2114). Nothing in the system variables mentions this last `.1` part though, so it is still possible to have both JDKs installed and select one at random once again.

To sum this up, we can have multiple JDKs with the same `java.version` and even the same `java.runtime.version` installed alongside and it is too hard/impossible for us to deduce which one is "newer". I think we should either use the JDK's installation path as a tie-breaker to at least achieve determinism or throw an exception to warn the user. As a user I tend to prefer the latter option because the installation path is arbitrary and its significance may be surprising.

Reply:
Should we just sort these lexically vs trying to parse them+sort? Is the `java.version` sanely sortable in that way?

Reply:
Some parsing seems to be necessary anyway because, for example, `11.2.10` is lexically greater than `11.10.2`.

Reply:
yes, you're right. I don't know what I was thinking.